### PR TITLE
SageMaker on Flyte: TrainingJob for training with built-in algorithms and basic HPOJob support  [Alpha]

### PR DESCRIPTION
**Pull request by bnsblue**
# TL;DR
This PR adds the necessary definitions for basic support of SageMaker TrainingJob and HPOJob (Hyperparameter Optimization)

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [x] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
This PR adds the basic support for users to invoke SageMaker TrainingJob (built-in algorithm mode) and SageMaker HPOJob from within Flyte. The follow paragraphs demonstrates the supported use cases.

## Defining a Simple Training Job
Users can leverage SageMaker's powerful built-in algorithms easily without needing to write any function or logic. They can simply define a SdkBuildinAlgorithmTrainingJobTask in Flytekit and supplies the settings and the spec of the built-in algorithm as follows:
```python
alg_spec = training_job_models.AlgorithmSpecification(
    input_mode=_sdk_sagemaker_types.InputMode.FILE,
    algorithm_name=_sdk_sagemaker_types.AlgorithmName.XGBOOST,
    algorithm_version="0.72",
    metric_definitions=[training_job_models.MetricDefinition(name="Minimize", regex="validation:error")]
)

# Definition of a SageMaker training job using SageMaker's built-in algorithm mode of XGBoost
xgboost_simple_train_task = training_job_task.SdkBuiltinAlgorithmTrainingJobTask(
    training_job_config=training_job_models.TrainingJobConfig(
        instance_type="ml.m4.xlarge",
        instance_count=1,
        volume_size_in_gb=25,
    ),
    algorithm_specification=alg_spec,
    cache_version='v1',
    cacheable=True,
)
```

## Defining a simple Hyperparameter Tuning Job
SageMaker-on-Flyte also supports easy chaining between a TrainingJob task and a hpo job. After users define a TrainingJob task, he/she may want to apply hyperparameter tuning to the the training, while also maintaining the flexibility to run the TrainingJob task standalone. This should be easily doable by using the SdkSimpleHPOJobTask in our SDK. SdkSimpleHPOJobTask accepts the definition of a TrainingJob task as a part of the spec.

```python
# Definition of a SageMaker hyperparameter-tuning job.
# Note that is chained behind the simple TrainingJob task defined above
xgboost_simple_hpo_task = hpo_job_task.SdkSimpleHPOJobTask(
    training_job=xgboost_simple_train_task,
    max_number_of_training_jobs=10,
    max_parallel_training_jobs=5,
    cache_version='2',
    retries=2,
    cacheable=True,
)
```

## Invoking Training Jobs Task and Hyperparameter Tuning Jobs Task
Invoking Training Job Tasks and HPO Job Tasks from inside a Flyte workflow is pretty much the same as invoking other types of tasks. You should be able to achieve this by simply supplying the required inputs to the task definition. For Training Job Tasks and HPO Job Tasks , required inputs are pre-defined inputs that we think is needed for every training job. That's why you don't see the declaration of these inputs in the task definition -- we added them for you in our SDK.
```python
@workflow_class
class SageMakerSimpleWorkflow(object):
    static_hyperparameters = Input(Types.Generic, required=True, help="Sample hyperparameter input")

    my_simple_trianing_task_exec = xgboost_simple_train_task(
        train="s3://path/to/train/data",
        validation="s3://path/to/validation/data",
        static_hyperparameters=static_hyperparameters,
        stopping_condition=StoppingCondition(
            max_runtime_in_seconds=43200,
        ).to_flyte_idl(),
    )
    ...
```
Flyte's Single Task Execution capability also makes it easy to invoke a SimpleTrainingJobTask and SimpleHPOJobTask. That is, users do not need a workflow to launch the SageMaker tasks; instead, they can simply define the tasks, and then register and launch the tasks standalone, which enables fast iterations.

```python
xgboost_simple_train_task = training_job_task.SdkBuiltinAlgorithmTrainingJobTask( ... )

xgboost_simple_hpo_task = hpo_job_task.SdkSimpleHPOJobTask(
    training_job=xgboost_simple_train_task,
    ...
)
```

## Tracking Issue
https://github.com/lyft/flyte/issues/255

## Follow-up issue
https://github.com/lyft/flyte/issues/431

### CODE REVIEW

File: flytekit/common/tasks/sagemaker_task.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+from flytekit.models import (
+    literals as _literal_models,
+    task as _task_models,
+    interface as _interface_model,
+)
+from google.protobuf.json_format import MessageToDict, ParseDict
+import gen.pb_python.sagemaker_pb2 as sagemaker_pb2
+from flytekit.common.exceptions import scopes as _exception_scopes
+
+_TASK_TYPE = "aws_sagemaker_hpo"
```
Comment (kumare3) [inline flytekit/common/tasks/sagemaker_task.py:16]: you want to use the declared constant?

File: flytekit/common/tasks/sagemaker_task.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+class SagemakerXgBoostOptimizer(_sdk_task.SdkTask):
```
Comment (kumare3) [inline flytekit/common/tasks/sagemaker_task.py:18]: rename this to HPO?

File: flytekit/common/tasks/sagemaker_task.py (hunk 1 above)
Comment (bnsblue) [reply flytekit/common/tasks/sagemaker_task.py:16]: Could you explain what you mean? I don't get the question

File: flytekit/common/tasks/sagemaker_task.py (hunk 2 above)
Comment (bnsblue) [reply flytekit/common/tasks/sagemaker_task.py:18]: I actually committed this file by mistake. This file hasn't played well with the rest of the definition

**Comment by bnsblue** [comment]
🔨

File: flytekit/models/sagemaker/hpo_job.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
+from __future__ import absolute_import
+
+from flyteidl.plugins.sagemaker import hyperparameter_tuning_job_pb2 as _pb2_hpo_job
+from flytekit.models import common as _common
+from flytekit.sdk.sagemaker import types as _sdk_sagemaker_types
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/hpo_job.py:5]: let's remove this.  model files are supposed to be 1:1 to their protobuf generated counterparts.  there should be no need to import anything else from flytekit.  take a look at some of the other ones to see how enums are handled for instance.

File: flytekit/models/sagemaker/hpo_job.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+from flytekit.common.exceptions import user as _user_exceptions
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/hpo_job.py:6]: should be able to remove this too.

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
+from __future__ import absolute_import
+
+from typing import Dict, List
+from flyteidl.plugins.sagemaker import parameter_ranges_pb2 as _idl_parameter_ranges
+from flytekit.models import common as _common
+from flytekit.sdk.sagemaker import types as _sdk_sagemaker_types
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/parameter_ranges.py:6]: same

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+
+class ParameterRanges(_common.FlyteIdlEntity):
+    def __init__(
+            self,
+            parameter_range_map: Dict[str, _common.FlyteIdlEntity],
+    ):
+        self._parameter_range_map = parameter_range_map
+
+    def to_flyte_idl(self):
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/parameter_ranges.py:210]: add type hint or type comment

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+                converted[k] = _idl_parameter_ranges.ParameterRangeOneOf(continuous_parameter_range=v.to_flyte_idl())
+            else:
+                converted[k] = _idl_parameter_ranges.ParameterRangeOneOf(categorical_parameter_range=v.to_flyte_idl())
+
+        return _idl_parameter_ranges.ParameterRanges(
+            parameter_range_map=converted,
+        )
+
+    @classmethod
+    def from_flyte_idl(cls, pb2_object):
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/parameter_ranges.py:225]: add type hint or type comment

File: flytekit/models/sagemaker/parameter_ranges.py (hunk 6 above)
Comment (wild-endeavor) [reply flytekit/models/sagemaker/parameter_ranges.py:210]: If you `from __future__ import annotations` at the top of the file, you can do forward references even in python 3.6.  That is, you can use the class as a type hint in the class definition itself.

See https://www.python.org/dev/peps/pep-0563/ for more info.

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    def scaling_type(self):
+        """
+        enum value from HyperparameterScalingType
+        :return: int
+        """
+        return self._scaling_type
+
+    def to_flyte_idl(self):
+        """
+        :return: _idl_parameter_ranges.IntegerParameterRange
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/parameter_ranges.py:132]: if using type comments, let's stick with `rtype` since that's the pattern matt established.

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+            raise _user_exceptions.FlyteValidationException("Invalid SageMaker HyperparameterScalingType specified")
+
+        return _idl_parameter_ranges.IntegerParameterRange(
+            max_value=self._max_value,
+            min_value=self._min_value,
+            scaling_type=scaling_type,
+        )
+
+    @classmethod
+    def from_flyte_idl(cls, pb2_object):
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/parameter_ranges.py:153]: type hint/type comment

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+            raise _user_exceptions.FlyteValidationException("Invalid SageMaker HyperparameterScalingType specified")
+
+        return _idl_parameter_ranges.ContinuousParameterRange(
+            max_value=self._max_value,
+            min_value=self._min_value,
+            scaling_type=scaling_type,
+        )
+
+    @classmethod
+    def from_flyte_idl(cls, pb2_object):
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/parameter_ranges.py:74]: type hint or type comment

File: flytekit/models/sagemaker/parameter_ranges.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    def scaling_type(self):
+        """
+        enum value from HyperparameterScalingType
+        :return: int
+        """
+        return self._scaling_type
+
+    def to_flyte_idl(self):
+        """
+        :return: _idl_parameter_ranges.ContinuousParameterRange
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/parameter_ranges.py:53]: `rtype`

File: flytekit/models/sagemaker/hpo_job.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    @property
+    def max_number_of_training_jobs(self):
+        return self._max_number_of_training_jobs
+
+    @property
+    def max_parallel_training_jobs(self):
+        return self._max_parallel_training_jobs
+
+    def to_flyte_idl(self):
+        return _pb2_hpo_job.HyperparameterTuningJob(
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/hpo_job.py:146]: type hint/type comment

File: flytekit/models/sagemaker/hpo_job.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+            max_number_of_training_jobs=self._max_number_of_training_jobs,
+            max_parallel_training_jobs=self._max_parallel_training_jobs,
+            training_job=self._training_job.to_flyte_idl(),  # SDK task has already serialized it
+        )
+
+    @classmethod
+    def from_flyte_idl(cls, pb2_object):
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/hpo_job.py:153]: type hint/type comment

File: flytekit/models/sagemaker/training_job.py [hunk 14]
```diff
@@ -0,0 +1,247 @@
+from __future__ import absolute_import
+
+from typing import List
+from flyteidl.plugins.sagemaker import training_job_pb2 as _training_job
+from flytekit.models import common as _common
+from flytekit.sdk.sagemaker import types as _sdk_sagemaker_types
```
Comment (wild-endeavor) [inline flytekit/models/sagemaker/training_job.py:6]: nix

File: flytekit/sdk/sagemaker/types.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
+import enum
+
+
+class HyperparameterScalingType(enum.Enum):
```
Comment (wild-endeavor) [inline flytekit/sdk/sagemaker/types.py:4]: These should all be in the models directory.  Take a look at the `class ResourceType`

File: sample-notebooks/All-hands Demo.ipynb [hunk 16]
```diff
@@ -0,0 +1,494 @@
...
+    "    \"validation\": \"s3://lyft-modelbuilder/test-datasets/pima/validation\",\n",
+    "    \"static_hyperparameters\": xgboost_hyperparameters,\n",
+    "    \"stopping_condition\": StoppingCondition(\n",
+    "        max_runtime_in_seconds=43200,\n",
+    "    ).to_flyte_idl(),\n",
+    "}\n",
+    "\n",
+    "\n",
+    "# A failed sagemaker task example\n",
+    "# https://flyte-staging.lyft.net/console/projects/flyteexamples/domains/development/executions/tlomtoafvh"
```
Comment (wild-endeavor) [inline sample-notebooks/All-hands Demo.ipynb:222]: @EngHabu not a huge fan of having these out in the public.   what do you think?

File: sample-notebooks/All-hands Demo.ipynb [hunk 17]
```diff
@@ -0,0 +1,494 @@
...
+    "        volume_size_in_gb=25,\n",
+    "    ),\n",
+    "    algorithm_specification=alg_spec,\n",
+    "    cache_version='3',\n",
+    "    cacheable=True,\n",
+    ")\n",
+    "\n",
+    "# Defining inputs for the TrainingJob task\n",
+    "training_inputs={\n",
+    "    \"train\": \"s3://lyft-modelbuilder/test-datasets/pima/train\",\n",
```
Comment (wild-endeavor) [inline sample-notebooks/All-hands Demo.ipynb:212]: @EngHabu not a huge fan of having these out in the public. what do you think?

File: tests/flytekit/common/workflows/sagemaker.py [hunk 18]
```diff
@@ -0,0 +1,27 @@
+from __future__ import absolute_import
```
Comment (wild-endeavor) [inline tests/flytekit/common/workflows/sagemaker.py:1]: is this file supposed to be here?

File: setup.py [hunk 19]
```diff
@@ -30,7 +30,7 @@
         ]
     },
     install_requires=[
-        "flyteidl>=0.17.34,<1.0.0",
+        "flyteidl==0.18.0b5",
```
Comment (EngHabu) [inline setup.py:33]: Reminder here to revert this line when done

File: sample-notebooks/All-hands Demo.ipynb (hunk 16 above)
Comment (EngHabu) [reply sample-notebooks/All-hands Demo.ipynb:222]: yeah ChangHong will take all of these out...

I think we need to think of how to demonstrate things that won't work on Sandbox in cookbook... generally speaking

File: sample-notebooks/All-hands Demo.ipynb (hunk 17 above)
Comment (EngHabu) [reply sample-notebooks/All-hands Demo.ipynb:212]: agreed

File: flytekit/models/sagemaker/hpo_job.py (hunk 3 above)
Comment (bnsblue) [reply flytekit/models/sagemaker/hpo_job.py:5]: SparkTask did the same tho (I actually referenced this when I wrote mine) https://github.com/lyft/flytekit/blob/a106ba7c3946837039b807d14f62fd2cc7d7c53b/flytekit/models/task.py#L616

Comment (bnsblue) [reply flytekit/models/sagemaker/hpo_job.py:5]: Also, when users write code they need to have these enums available in sdk. I think having this is reasonable

File: flytekit/plugins/__init__.py [hunk 20]
```diff
@@ -34,7 +34,7 @@
 _lazy_loader.LazyLoadPlugin(
     "schema",
     [
-        "numpy>=1.14.0,<2.0.0",
+        "numpy>=1.14.0,<1.19.0",
```
Comment (EngHabu) [inline flytekit/plugins/__init__.py:37]: Please revert this. This PR has been approved: https://github.com/lyft/flytekit/pull/144

File: flytekit/models/sagemaker/hpo_job.py (hunk 3 above)
Comment (bnsblue) [reply flytekit/models/sagemaker/hpo_job.py:5]: @wild-endeavor 
I'll move it to models, is the expectation for users to access the models wrapper? I feel that's a bit awkward, what do you think?

File: sample-notebooks/All-hands Demo.ipynb (hunk 16 above)
Comment (bnsblue) [reply sample-notebooks/All-hands Demo.ipynb]: I moved them to my github enterprise repo for now. Let's figure out a long term place for them

File: flytekit/models/sagemaker/parameter_ranges.py (hunk 6 above)
Comment (bnsblue) [reply flytekit/models/sagemaker/parameter_ranges.py]: Since our CI is using 3.6.3 and `from __future__ import annotations` is only supported starting python 3.7, I'll remove this import and remove the type hints for `self`s. Thanks for the suggestion tho.

File: flytekit/common/tasks/sagemaker/training_job_task.py [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+from flytekit.common import interface as _interface
+from flytekit.models.sagemaker import training_job as _training_job_models
+from google.protobuf.json_format import MessageToDict
+from flytekit.models import types as _idl_types
+from flytekit.models.core import types as _core_types
+from flytekit.models import literals as _literal_models
+from flytekit.common.constants import SdkTaskType
+
+
+class SdkSimpleTrainingJobTask(_sdk_task.SdkTask):
```
Comment (kumare3) [inline flytekit/common/tasks/sagemaker/training_job_task.py:19]: What does Simple mean?

File: flytekit/common/tasks/sagemaker/training_job_task.py [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+            interface=_interface.TypedInterface(
+                inputs={
+                    "static_hyperparameters": _interface_model.Variable(
+                        type=_idl_types.LiteralType(simple=_idl_types.SimpleType.STRUCT),
+                        description="",
+                    ),
+                    "train": _interface_model.Variable(
+                        type=_idl_types.LiteralType(
+                            blob=_core_types.BlobType(
+                                format="csv",
```
Comment (kumare3) [inline flytekit/common/tasks/sagemaker/training_job_task.py:70]: is this always CSV?

File: flytekit/common/tasks/sagemaker/training_job_task.py (hunk 21 above)
Comment (kumare3) [reply flytekit/common/tasks/sagemaker/training_job_task.py:19]: is this meant to be built in?

File: flytekit/common/tasks/sagemaker/training_job_task.py [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+        self.add_inputs(
+            {
+                "static_hyperparameters": _interface_model.Variable(
+                    type=_idl_types.LiteralType(simple=_idl_types.SimpleType.STRUCT),
+                    description="",
+                ),
+                "train": _interface_model.Variable(
+                    type=_idl_types.LiteralType(
+                        blob=_core_types.BlobType(
+                            format="csv",
```
Comment (kumare3) [inline flytekit/common/tasks/sagemaker/training_job_task.py:170]: same thing, I think we should take the "data-format" and default it to csv. We should provide the options (enum)? Currently in the Enum you can have only csv

File: flytekit/common/tasks/sagemaker/training_job_task.py (hunk 21 above)
Comment (bnsblue) [reply flytekit/common/tasks/sagemaker/training_job_task.py:19]: Yes -- meaning users don't write his/her own decorated function.

File: flytekit/models/sagemaker/training_job.py [hunk 24]
```diff
@@ -0,0 +1,231 @@
...
+    @property
+    def name(self) -> str:
+        """
+
+        :rtype: str
+        """
+        return self._name
+
+    @property
+    def regex(self) -> str:
```
Comment (kumare3) [inline flytekit/models/sagemaker/training_job.py:83]: we might need to add documentation here :(. This is the sad part of models. You can do that maybe?

File: flytekit/models/sagemaker/training_job.py [hunk 25]
```diff
@@ -0,0 +1,231 @@
...
+        return _training_job_pb2.MetricDefinition(
+            name=self.name,
+            regex=self.regex,
+        )
+
+    @classmethod
+    def from_flyte_idl(cls, pb2_object: _training_job_pb2.MetricDefinition):
+        """
+
+        :param pb2_object: _training_job_pb2.MetricDefinition
```
Comment (kumare3) [inline flytekit/models/sagemaker/training_job.py:104]: I dont think you need this? Why not add the return type as well `-> ...`

Comment (kumare3) [reply flytekit/models/sagemaker/training_job.py:104]: Nevermind about the return type, yes we cannot because 3.6 does not support forward references

Comment (bnsblue) [reply flytekit/models/sagemaker/training_job.py:104]: exactly

**Comment by kumare3** [comment]
This looks awesome, two major comments
1. Support for other formats (currently you can keep it at CSV)
2. Docs to be added

**Comment by kumare3** [comment]
Looks good to me, but we should get plugin merged before this

File: tests/flytekit/unit/sdk/tasks/test_sagemaker_tasks.py [hunk 26]
```diff
@@ -129,6 +130,8 @@ def test_simple_training_job_task():
 
 
 def test_simple_hpo_job_task():
+    print(simple_xgboost_hpo_job_task.custom["trainingJob"])
```
Comment (wild-endeavor) [inline tests/flytekit/unit/sdk/tasks/test_sagemaker_tasks.py:133]: nix or use logger?
